### ISSUE TITLE
Web UI login error after update to v0.19.2.0 -- "Request Failed With Status Code 500"

### ISSUE THREAD

**Issue by quaintyurt** [slskd/slskd#1019]
I am running in a Docker (v20.10.24) container on unRaid (v6.12.4). After upgrading the container I am no longer able to authenticate with the webUI. The logs show an error stating incorrect KeyedHashAlgorithm for HS256 (key size = 176bits; expected size = 256bits).

Other container versions:
.NET 7.0.14
Unix 6.1.49.0

![slskd - Vivaldi -- 5iO](https://github.com/slskd/slskd/assets/152746204/8abbdf71-df0e-44f7-85e3-491ed1bb9e92)

```12/17/2023
12:06:25 PM
System.ArgumentOutOfRangeException: IDX10720: Unable to create KeyedHashAlgorithm for algorithm 'HS256', the key size must be greater than: '256' bits, key has '176' bits. (Parameter 'keyBytes')
12/17/2023
12:06:25 PM
   at Microsoft.IdentityModel.Tokens.CryptoProviderFactory.ValidateKeySize(Byte[] keyBytes, String algorithm, Int32 expectedNumberOfBytes)
12/17/2023
12:06:25 PM
   at Microsoft.IdentityModel.Tokens.CryptoProviderFactory.CreateKeyedHashAlgorithm(Byte[] keyBytes, String algorithm)
12/17/2023
12:06:25 PM
   at Microsoft.IdentityModel.Tokens.SymmetricSignatureProvider.CreateKeyedHashAlgorithm()
12/17/2023
12:06:25 PM
   at Microsoft.IdentityModel.Tokens.DisposableObjectPool`1.CreateInstance()
12/17/2023
12:06:25 PM
   at Microsoft.IdentityModel.Tokens.DisposableObjectPool`1.Allocate()
12/17/2023
12:06:25 PM
   at Microsoft.IdentityModel.Tokens.SymmetricSignatureProvider.GetKeyedHashAlgorithm(Byte[] keyBytes, String algorithm)
12/17/2023
12:06:25 PM
   at Microsoft.IdentityModel.Tokens.SymmetricSignatureProvider.Sign(Byte[] input)
12/17/2023
12:06:25 PM
   at Microsoft.IdentityModel.JsonWebTokens.JwtTokenUtilities.CreateEncodedSignature(String input, SigningCredentials signingCredentials)
12/17/2023
12:06:25 PM
   at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.WriteToken(SecurityToken token)
12/17/2023
12:06:25 PM
   at slskd.Core.API.TokenResponse.get_Token()
12/17/2023
12:06:25 PM
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.GetMemberAndWriteJson(Object obj, WriteStack& state, Utf8JsonWriter writer)
12/17/2023
12:06:25 PM
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryWrite(Utf8JsonWriter writer, T value, JsonSerializerOptions options, WriteStack& state)
12/17/2023
12:06:25 PM
   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
12/17/2023
12:06:25 PM
   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
12/17/2023
12:06:25 PM
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.SerializeAsync(Stream utf8Json, T rootValue, CancellationToken cancellationToken, Object rootValueBoxed)
12/17/2023
12:06:25 PM
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.SerializeAsync(Stream utf8Json, T rootValue, CancellationToken cancellationToken, Object rootValueBoxed)
12/17/2023
12:06:25 PM
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.SerializeAsync(Stream utf8Json, T rootValue, CancellationToken cancellationToken, Object rootValueBoxed)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Mvc.Formatters.SystemTextJsonOutputFormatter.WriteResponseBodyAsync(OutputFormatterWriteContext context, Encoding selectedEncoding)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|30_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultFilters()
12/17/2023
12:06:25 PM
--- End of stack trace from previous location ---
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
12/17/2023
12:06:25 PM
--- End of stack trace from previous location ---
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
12/17/2023
12:06:25 PM
   at slskd.HTMLInjectionMiddleware.InvokeAsync(HttpContext context)
12/17/2023
12:06:25 PM
   at slskd.HTMLRewriteMiddleware.InvokeAsync(HttpContext context)
12/17/2023
12:06:25 PM
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddlewareImpl.<Invoke>g__Awaited|8_0(ExceptionHandlerMiddlewareImpl middleware, HttpContext context, Task task)```

**Comment by jpdillingham** [comment]
Thanks for reporting this.  Whatever the issue is would have been introduced in #998, probably [this line](https://github.com/slskd/slskd/pull/998/files#diff-6771e8d1fae4904ab288c5069d124e68596e3aeab896d12307e50ba719a0cf83R76) which bumped `System.IdentityModel.Tokens.Jwt` from 6.25.1 to 7.0.3.

The exception indicates that the configured JWT key isn't long enough.  I was able to reproduce the problem on my end by shortening my key via the UI, which I shouldn't have been able to do if validation was configured properly.

The quick fix will be to downgrade the container to 0.19.1.  I'll be working on figuring out what changed in the underlying library that caused this breakage.

**Comment by jpdillingham** [comment]
Refreshing myself on the configuration [here](https://github.com/slskd/slskd/blob/master/docs/config.md#authentication), the JWT key is expected to have a length between 16 and 255, and that's reflected in the validation [here](https://github.com/slskd/slskd/blob/master/src/slskd/Core/Options.cs#L1653)

This key is used to create an instance of `SymmetricSecurityKey` [here](https://github.com/slskd/slskd/blob/master/src/slskd/Program.cs#L630) and in turn that instance is used to sign generated JWTs [here](https://github.com/slskd/slskd/blob/master/src/slskd/Core/Security/SecurityService.cs#L88)

Given this it appears as though the upgrade of `System.IdentityModel.Tokens.Jwt` from major version 6 to 7 introduced a constraint requiring the supplied key to be at least 256 bits.

I'll see if I can find any evidence to back this up.  I think the fix will be to pad the configured key with spaces to ensure it meets minimum length.

**Comment by jpdillingham** [comment]
That is indeed the case.  See: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2072

**Comment by jpdillingham** [comment]
Fix should be out ~20 minutes from this post.

I increased the length of the default signing key from 16 to 32 characters.  This should fix the problem for anyone that hasn't specified a custom signing key.

For those that have specified a custom key, I added logic to pad the length of the key out to 32 characters, and added a warning:

```
[slskd.Program] [16:37:17 WRN] The configured JWT signing key is less than the required 32 characters and has been padded with spaces to make up the difference. The minimum length will be raised from 16 to 32 in a future release, and this will become an error.
```

Note that this will break interoperability for anyone that would be using the key for more than just slskd (this should be an extreme edge case), and they'll have to adjust accordingly.  In the future I'll increase the minimum length for the key to 32 characters, which will probably be a breaking change for anyone that has configured a custom key.
